### ISSUE TITLE
Error: 'b' format requires -128 <= number <= 127

### ISSUE THREAD

**Issue by WallisNi** [h4de5/home-assistant-toshiba_ac#208]
Hi! 

Firstly, thank you for all the effort put into this! 

Now, since a few days, don't know when for sure, I'm getting the following error message for any attempt setting of my device, haori.

Error: 'b' format requires -128 <= number <= 127

Is there anything I can do to resolve this? Reinstall is not helping.

**Comment by WallisNi** [comment]
So, I opened the Toshiba Android app and agreed to the new terms. It seems to have resolved the issue. Will update if it reappears.

**Comment by WallisNi** [comment]
And updating again. The integration failed again. But after opening the official app and instantly returning to make changes in HA the setting changed.

**Comment by cuetus** [comment]
Same problem. Also reloading the integration solves the issue for some time, then it returns with no clear pattern.

**Comment by WallisNi** [comment]
I've been okay a couple of days now. But still unsure what caused the initial issue. Will report back if it reappears.

**Comment by cuetus** [comment]
I started to reload the integration every morning but there really is no pattern. Has worked ok until now - right now it has failed to set the temperature again with the same error (setting temperatures based on electricity spot prices every hour)

**Comment by cuetus** [comment]
I put integration reload to happen every morning at 3:50AM, but this morning it has failed again. This same error coming from every function, does not matter if I try to set fan speed, temperature, operating mode or what. But again, reloading the integration fixed the issue... until next time.

**Comment by cuetus** [comment]
The way to use this now is in automations by first reloading the integration and then doing what you need to do, this same every time you want to change something. Don't know if this is specific to just Toshiba Polar 35 with adapter version 2.4.00.

**Comment by Arno-MA-73** [comment]
I have the same problem. Home Assistant 2024.11.2, adapter version 2.4.00.  We have a Toshiba multi split system with two indoor units: a Premium 13+ model and a floor model.

I have used this Toshiba integration for almost a year, and the last couple of days it has become unusable.

**Comment by Arno-MA-73** [comment]
Here is the log entry I see: 

Logger: homeassistant.components.automation.turn_off_heat_pump_when_warm
Source: components/automation/__init__.py:743
integration: Automation (documentation, issues)
First occurred: 09:49:26 (4 occurrences)
Last logged: 20:49:23

While executing automation automation.turn_off_heat_pump_when_warm
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/automation/__init__.py", line 743, in async_trigger
    return await self.action_script.async_run(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1801, in async_run
    return await asyncio.shield(create_eager_task(run.async_run()))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 464, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 528, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 558, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 526, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 781, in _async_device_step
    await device_action.async_call_action_from_config(
  File "/usr/src/homeassistant/homeassistant/components/device_automation/action.py", line 72, in async_call_action_from_config
    await platform.async_call_action_from_config(hass, config, variables, context)
  File "/usr/src/homeassistant/homeassistant/components/climate/device_action.py", line 98, in async_call_action_from_config
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 2802, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2845, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1007, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1079, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 772, in async_handle_set_hvac_mode_service
    await self.async_set_hvac_mode(hvac_mode)
  File "/config/custom_components/toshiba_ac/climate.py", line 194, in async_set_hvac_mode
    await self._device.set_ac_status(ToshibaAcStatus.OFF)
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 246, in set_ac_status
    await self.send_state_to_ac(state)
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 147, in send_state_to_ac
    future_state = ToshibaAcFcuState.from_hex_state(self.fcu_state.encode())
                                                    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/fcu_state.py", line 281, in encode
    encoded = self.ENCODING_STRUCT.pack(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: 'b' format requires -128 <= number <= 127

**Comment by pjylkka** [comment]
Yes, this is becoming hurtful.  Rebooting HA will fix it for sometime, but after half hour/hour or so - changing for example temperature becomes impossible and automations won't work due of:
```
[140061020839296] Unexpected exception
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 245, in handle_call_service
    response = await hass.services.async_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2802, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2845, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1007, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1079, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 1029, in async_service_temperature_set
    await entity.async_set_temperature(**kwargs)
  File "/config/custom_components/toshiba_ac/climate.py", line 120, in async_set_temperature
    await self._device.set_ac_temperature(set_temperature)
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 275, in set_ac_temperature
    await self.send_state_to_ac(state)
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/__init__.py", line 147, in send_state_to_ac
    future_state = ToshibaAcFcuState.from_hex_state(self.fcu_state.encode())
                                                    ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/toshiba_ac/device/fcu_state.py", line 281, in encode
    encoded = self.ENCODING_STRUCT.pack(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: 'b' format requires -128 <= number <= 127
```

**Comment by AtzeDeVries** [comment]
same issue here as well...

**Comment by wiblaa** [comment]
Same issue here... making a change in the Toshiba app on the phone helps for a while, or until sufficient commands have been sent from HA, then it breaks again.

**Comment by DjLysy** [comment]
It just happened to me after accepting new terma in Toshiba mobile app. After that, i have reloaded the HA Toshiba integration plugin and started to work again correctly

**Comment by wiblaa** [comment]
> It just happened to me after accepting new terma in Toshiba mobile app. After that, i have reloaded the HA Toshiba integration plugin and started to work again correctly

Does it work after a few hours?

**Comment by cuetus** [comment]
![Screenshot 2024-11-22 at 12 25 41](https://github.com/user-attachments/assets/d24698e0-009a-43cf-ad47-21b34ea40aab)
No, the terms and conditions have been accepted etc but it still fails, in the screenshot 30 minutes ago the last time when the check was run. I run this like this now, using node red instead of HA automations, my automations run at every hour, so 1 minute before I get the status out, if that works I set the fan mode to whatever the status returned (=no change) and if either of these nodes fail I reload the integration. So.. that is a workaround that works for me.

**Comment by AtzeDeVries** [comment]
> It just happened to me after accepting new terma in Toshiba mobile app. After that, i have reloaded the HA Toshiba integration plugin and started to work again correctly

This works only for a limited time. Reloading the integration fixes this, but for a limited time again.

**Comment by johnbrady833** [comment]
Thanks @pjylkka, I raised #211 and I think you're right, that's a duplicate of this issue.

I found that rolling back from HA 2024.11.2 to an earlier version fixed it, so I assume something has changed in 2024.11.2 that is incompatible with this integration. Someone mentioned on the other thread that there was a change in the Websocket implementation which had caused other plugins to break.

**Comment by wiblaa** [comment]
I rolled back to:
Core 2024.10.4
Supervisor 2024.11.4
... and still see the same issue. Weirdness.

**Comment by johnbrady833** [comment]
@wiblaa I'm now seeing the same issue. After opening the Toshiba app and clicking "Agree", the integration doesn't work for me either.

**Comment by psoininen** [comment]
Same issue, probably appeared already a couple days ago with some automation acting up but only noticed the same error message trying to set fan manually now: 'b' format requires -128 <= number <= 127. Have not updated core beyond 10.4.

**Comment by WallisNi** [comment]
It seems to me a reinstall of the plugin helps. I've been good for several
days in between. But today I installed the latest ha update, and lost the
heater. It seems to work as expected after a reinstall from hacs.

**Comment by WallisNi** [comment]
Error seems to be present att about 3/4 of commands. Can't figure out if there are other circumstances causing it. Looking into using Google assistant instead until plugin is updated. Not ideal, but at least commands seem to get sent that way.

**Comment by elsonico** [comment]
I have same issue. Even when it works little while after restart this totally breaks my automation which does not work anymore due this.

**Comment by cuetus** [comment]
Toshiba must be changing something. Integration has started to go 'unavailable' after a while now, need to test for that status as well instead of just catching the exceptions in the reloading health checks. Reloading the integration fixes that as well, until it breaks again.

**Comment by MickeBergman** [comment]
I also have this since some time. It comes and goes by random.
And even more confusing but might help solving is that I have 2 heat pumps and one works and the other not.
Now I tried and the "non-working" accepted commands from the Toshiba app. And HA immediately showed the changes. AND worked for one command and then back to the error.

**Comment by pbock** [comment]
This seems to be an issue with the [toshiba-ac](https://github.com/KaSroka/Toshiba-AC-control) package where I’ve submitted [an issue](https://github.com/KaSroka/Toshiba-AC-control/issues/61). I have a suspicion that this only happens when outside temperatures are below zero, does that seem right?

**Comment by wiblaa** [comment]
We droppe below freezing a few days ago, so that tracks...

**Comment by MickeBergman** [comment]
This actually could be. One of mine is -5 and gives the error - other +6 and works.

**Comment by WallisNi** [comment]
Seems to be the case for me also. Don't have the logs to verify, but the automation traces for yesterday and today are ok, and we've been above zero during that time. Nice catch @pbock !

**Comment by cuetus** [comment]
<img width="1191" alt="Screenshot 2024-11-26 at 14 04 01" src="https://github.com/user-attachments/assets/189a8079-0133-4bbe-b958-4ad43b8bd156">
Yep, we have been below 0 last month.

**Comment by narfotic** [comment]
Having issues as well. Not below zero at the moment.

**Comment by Arno-MA-73** [comment]
@h4de5 could you have a look at this issue please? Temperatures will be below zero next week and I suspect that my Toshiba automations will break then again.

**Comment by h4de5** [comment]
as you have already stated, this error is coming from the api package below. the integration here is just a shell for that. this is the same for most of the other issues here too - there is hardly anything i can change.

**Comment by stianhaugli** [comment]
+1 from Norway. having the same issue when outdoor temp drops below zero. it seems 
[pbock](https://github.com/h4de5/home-assistant-toshiba_ac/issues/208#issuecomment-2499018046) is quite right. [kaSroka](https://github.com/KaSroka/Toshiba-AC-control/issues/61#issuecomment-2521214159) has just replied saying it'll be fixed, hopefully soon. thank you pbock (!)

**Comment by superior99** [comment]
same issue here as well with (toshiba v[2024.9.0] and latest HA.
I get this error:
"cannot perform the action climate/set_temperature . 'b' format requires -128 <= number <= 127"

Opening Toshiba app and/or restarting HA solves the problem for a while, then it starts to give errors again shortly after that.

**Comment by pbock** [comment]
Should be fixed in toshiba-ac v0.3.10 which has just been released! Any chance we could get this updated, @h4de5?

**Comment by h4de5** [comment]
fixed in #214
